File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# its the last in the level
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        q = deque()
        q.append(root)
        level = [root]
        res = []
        while q:
            level = list(q)
            for element in level:
                item = q.popleft()
                if item is None: continue
                if item.left: q.append(item.left)
                if item.right: q.append(item.right)
            
            if level[-1] is not None: res.append(level[-1].val)

        return res

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

/*
we do bfs and return the last element of the level
*/
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;
impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut res: Vec<i32> = vec![];
        let mut q: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        if let Some(r) = root {q.push_back(r)}

        while q.len() > 0 {
            let mut curr_level: Vec<i32> = vec![];
            let n = q.len();
            for i in 0..n {
                if let Some(node) = q.pop_front() {
                    let node_aux = node.borrow();
                    curr_level.push(node_aux.val);
                    if let Some(l) = node_aux.left.clone(){
                        q.push_back(l)
                    }
                    if let Some(r) = node_aux.right.clone(){
                        q.push_back(r)
                    }
                };
            }
            res.push(*curr_level.last().unwrap());
        } 
        return res; 
    }
}